File: Data Structures & Algorithms/car-fleet/submission-1.js
class Solution {
    /**
     * @param {number} target
     * @param {number[]} position
     * @param {number[]} speed
     * @return {number}
     */
    carFleet(target, position, speed) {
        const positionSpeedMap = {}

        for (let indx = 0; indx < position.length; indx++) {
            positionSpeedMap[position[indx]] = speed[indx]
        }

        position = position.sort((a, b) => b - a)
        const stack = []

        if (position.length <= 1) return position.length

        for (const carPosition of position) {
            const destReachTime = (target - carPosition) / positionSpeedMap[carPosition]
            stack.push(destReachTime)
            if (stack.length >= 2 && stack[stack.length - 1] <= stack[stack.length - 2]) {
                stack.pop()
            }
        }
        return stack.length
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-1.java
class Solution {
    int[] cache;

    public int climbStairs(int n) {
        cache = new int[n];
        for (int i = 0; i < n; i++) {
            cache[i] = -1;
        }
        return dfs(0, n);
    }

    public int dfs(int i, int target) {
        if (i == target)
            return 1;

        if (i > target)
            return 0;

        if (cache[i] != -1)
            return cache[i];

        cache[i] = dfs(i + 1, target) + dfs(i + 2, target);
        return cache[i];
    }
}
